composite/outlier-detection/src/adapter.ts

```diff
@@ -0,0 +1,93 @@
+import { AdapterRequest, AdapterResponse, Execute } from '@chainlink/types'
+import { Requester, Validator } from '@chainlink/external-adapter'
+import { getSourceImpl } from './source'
+import { getLatestAnswer } from '@chainlink/reference-data-reader'
+import { getCheckImpl } from './check'
+import { Config, makeConfig } from './config'
+
+const customParams = {
+  referenceContract: ['referenceContract', 'contract'],
+  multiply: true,
+}
+
+export const makeExecute = (config?: Config): Execute => {
+  return async (request: AdapterRequest) => execute(request, config || makeConfig())
+}
+
+export const execute = async (input: AdapterRequest, config: Config): Promise<AdapterResponse> => {
+  const sourceExecute = config.sourceDataProviders.map(getSourceImpl)
+  if (sourceExecute.length === 0) {
+    throw Error('No source adapters provided')
+  }
+  const checkExecute = config.checkDataProviders.map(getCheckImpl)
+  return await executeWithAdapters(input, sourceExecute, checkExecute, config)
+}
+
+export const executeWithAdapters = async (
+  input: AdapterRequest,
+  sources: Execute[],
+  checks: Execute[],
+  config: Config,
+): Promise<AdapterResponse> => {
+  const validator = new Validator(input, customParams)
+  if (validator.error) throw validator.error
+
+  if (!config.threshold) {
+    throw new Error('config is missing threshold values')
+  }
+
+  const jobRunID = validator.validated.id
+  const referenceContract = validator.validated.data.referenceContract
+  const multiply = validator.validated.data.multiply
+
+  const onchainValue = await getLatestAnswer(referenceContract, multiply, input.meta)
+
+  const sourceMedian = await getExecuteMedian(sources, input)
+  if (
+    config.threshold.onchain > 0 &&
+    difference(sourceMedian, onchainValue) > config.threshold.onchain
+  ) {
+    return returnValue(jobRunID, onchainValue)
+  }
+
+  if (config.threshold.checks > 0) {
+    if (checks.length === 0) {
+      throw Error('No check adapters provided')
+    }
+
+    const checkMedian = await getExecuteMedian(checks, input)
+    if (difference(sourceMedian, checkMedian) > config.threshold.checks) {
+      return returnValue(jobRunID, onchainValue)
+    }
+  }
+
+  return returnValue(jobRunID, sourceMedian)
+}
+
+const getExecuteMedian = async (executes: Execute[], request: AdapterRequest): Promise<number> => {
+  const responses = await Promise.allSettled(executes.map((execute) => execute(request)))
+  const values = responses
+    .filter((result) => result.status === 'fulfilled' && 'value' in result)
```

Comment:
Since that's how we do things with FM, I think this is the best approach to take. We also don't want the entire thing to fail if 1 out of 2 data providers fail.